File: src/rendering/managers/render.rs
use glyphon::{Resolution, TextArea, TextBounds};
use wgpu::*;

use crate::{engine_runtime::EngineRuntime, typed_addr::dupe};
//use crate::span;
//use crate::drop_span;

use super::RenderManagerResource;

impl EngineRuntime {
    pub fn render(&mut self) -> Result<(), SurfaceError> {
        let state = self.get_resource_mut::<RenderManagerResource>();

        dupe(&state.text_state)
            .text_render
            .prepare(
                &state.device,
                &state.queue,
                dupe(&state.text_state).font_system,
                dupe(&state.text_state).atlas,
                Resolution {
                    width: state.size.width,
                    height: state.size.height,
                },
                [TextArea {
                    buffer: &dupe(&state.text_state).buffer,
                    left: 10.0,
                    top: 10.0,
                    scale: 1.0,
                    bounds: TextBounds {
                        left: 0,
                        top: 0,
                        right: 600,
                        bottom: 160,
                    },
                    default_color: glyphon::Color::rgb(255, 255, 255),
                }],
                &mut state.text_state.cahce,
            )
            .unwrap();

        //span!(trace_get_surface_texture, "get surface texture");
        let output = state.surface.get_current_texture()?;
        //drop_span!(trace_get_surface_texture);

        let view = output
            .texture
            .create_view(&TextureViewDescriptor::default());

        let mut encoder = state
            .device
            .create_command_encoder(&CommandEncoderDescriptor {
                label: Some("Render encoder"),
            });

        //span!(trace_render_pass, "render pass"); // before this most of the time
        let mut render_pass = encoder.begin_render_pass(&RenderPassDescriptor {
            label: Some("Render Pass"),
            color_attachments: &[Some(RenderPassColorAttachment {
                view: &view,
                resolve_target: None,
                ops: Operations {
                    load: LoadOp::Clear(*state.clear_color),
                    store: StoreOp::Store,
                },
            })],
            depth_stencil_attachment: Some(RenderPassDepthStencilAttachment {
                view: &state.depth_texture.view,
                depth_ops: Some(Operations {
                    load: LoadOp::Clear(1.),
                    store: StoreOp::Store,
                }),
                stencil_ops: None,
            }),
            occlusion_query_set: None,
            timestamp_writes: None,
        });

        for draw_pass in dupe(state).draw_passes.iter_mut() {
            if draw_pass.is_loaded(self) {
                draw_pass.draw(&mut render_pass, dupe(self));
            }
        }

        drop(render_pass);
        //span!(trace_render_pass, "render pass"); // before this most of the time
        let mut render_pass = encoder.begin_render_pass(&RenderPassDescriptor {
            label: Some("Render Pass"),
            color_attachments: &[Some(RenderPassColorAttachment {
                view: &view,
                resolve_target: None,
                ops: Operations {
                    load: LoadOp::Load,
                    store: StoreOp::Store,
                },
            })],
            depth_stencil_attachment: None,
            occlusion_query_set: None,
            timestamp_writes: None,
        });

        state
            .text_state
            .text_render
            .render(state.text_state.atlas, &mut render_pass)
            .unwrap();
        drop(render_pass);

        state.draw_passes.clear();
        //span!(trace_light_material, "light_material");

        //drop_span!(trace_render_pass);
        //drop(render_pass);

        //span!(trace_queue, "queue submit");
        state.queue.submit(vec![encoder.finish()]);
        //drop_span!(trace_queue);

        //span!(trace_present, "present");
        output.present();
        //drop_span!(trace_present);

        state.text_state.atlas.trim();

        Ok(())
    }
}

File: krajc/src/rendering/systems/render.zig
const math = @import("../../math.zig");
const std = @import("std");
const wgpu = @import("wgpu");
const krajc = @import("../../prelude.zig");

pub const Vertex = struct {
    pos: [3]f32,
    p1: f32 = 0.0,
    color: [3]f32,
    p2: f32 = 0.0,

    pub fn new(pos: math.Vec3, color: math.Vec3) Vertex {
        return Vertex{
            .pos = [_]f32{ pos.x(), pos.y(), pos.z() },
            .color = [_]f32{ color.x(), color.y(), color.z() },
        };
    }
};

pub fn render(res: *krajc.ResourceState) !void {
    const state = res.get(krajc.RenderingState);

    const surface_view = state.surface_swap_chain.getCurrentTextureView().?;

    const color_attachment = wgpu.RenderPassColorAttachment{
        .view = surface_view,
        .resolve_target = null,
        .clear_value = std.mem.zeroes(wgpu.Color),
        .load_op = .clear,
        .store_op = .store,
    };
    const encoder = state.device.createCommandEncoder(&.{ .label = "Command Encoder" });

    const render_pass_info = wgpu.RenderPassDescriptor.init(.{
        .color_attachments = &.{color_attachment},
    });
    var render_pass = encoder.beginRenderPass(&render_pass_info);

    const vertex_buffer = state.device.createBuffer(&.{ .usage = .{ .vertex = false, .copy_dst = false }, .size = 512 * 512 });
    defer vertex_buffer.destroy();
    const vertex_data = [_]f32{
        //Vertex.new(math.Vec3.new(-1.0, -1.0, 0.0), math.Vec3.new(1.0, 1.0, 1.0)), // Bottom-left vertex
        //Vertex.new(math.Vec3.new(1.0, -1.0, 0.0), math.Vec3.new(1.0, 1.0, 1.0)), // Bottom-right vertex
        //Vertex.new(math.Vec3.new(0.0, 1.0, 0.0), math.Vec3.new(1.0, 1.0, 1.0)), // Top vertex
        1.0,
        2.0,
        3.0,
    };
    const slice = vertex_data[0..];
    state.queue.writeBuffer(
        vertex_buffer,
        0,
        slice,
    );

    render_pass.setPipeline(state.render_pipeline);
    render_pass.setVertexBuffer(0, vertex_buffer, 0, 4 * 3);
    render_pass.draw(3, 1, 0, 0);

    render_pass.end();
    render_pass.release();

    const command_buffer = encoder.finish(null);
    encoder.release();
    state.instance.processEvents();

    state.queue.submit(&[_]*wgpu.CommandBuffer{command_buffer});
    command_buffer.release();
    state.surface_swap_chain.present();
    surface_view.release();
    state.instance.processEvents();
}
